Voice Transformer Network

unilight opened the pull request:
Hi, I plan to open-source the Voice Transformer Network a.k.a. Transformer-VC.
Tl.dr: Parallel VC based on mel-spectrogram using TTS-pretrained Transformer (Tacotron 2 also supported). #1670 
https://arxiv.org/pdf/1912.06813.pdf

This is still WIP since I am still cleaning my local repo.
I am just opening this PR to ask you to help me check the design since I am touching some core functions (those in `espnet/`).

My original paper proposed pertaining with TTS (please check my paper), but I think it is troublesome to provide the script for pertaining. (Training from scratch does not reach satisfactory performance)
So I am thinking about providing the pretrained model (for training) only.

### Essential
- [x] Clean up `arctic/vc/run.sh`
- [x] Rename the main config files
- [x] Upload the pretrained model for training and write downloading scripts
- [x] Rename `vc` to `vc1`
- [x] write tests

### Documentation
- [x] Modify `egs/README.md`
- [x] Modify `README.md`
- [x] Modify `egs/arctic/README.md` to explain how to use this recipe?

### Misc
- [x] move `local/make_pair_json.py` to `utils`?
- [x] <del>Make symbolic links for some files in `conf/`?</del>

### Future work
- integration tests
- inherit the class from TTS ones and override some methods. (c.f., `e2e_asr_conformer`)

sw005320 commented:
One question.
If we do not specify the pre-trained model in `run.sh`, is it still working?

unilight commented:
> One question.
> If we do not specify the pre-trained model in `run.sh`, is it still working?

Yes. It is just that the quality will be greatly improved. 
https://unilight.github.io/Publication-Demos/publications/transformer-vc/
![image](https://user-images.githubusercontent.com/11516009/85308994-87ac9880-b4e4-11ea-8a9b-2d564718f791.png)

kan-bayashi commented:
`vc` -> `vc1`?

kan-bayashi commented:
Sorry for the late review. I will check in this evening.

kan-bayashi commented on egs/arctic/vc/local/ob_eval/evaluate.sh:
```diff
@@ -0,0 +1,152 @@
+        --gtwavdir ${db_root}/cmu_us_${trgspk}_arctic/wav \
+        --mcep_dim ${mcep_dim} \
+        --shiftms ${shift_ms} \
+        --f0min ${minf0} \
+        --f0max ${maxf0}
```
I have a general question about MCD.
In the case of world-based mcep, the MCD value is affected by the f0 prediction.
So I think the use of pysptk-based mcep is more robust.
How do you think?

unilight replied:
I have never calculated and compared pysptk-based mcep, but I have been calculated this way in my VTN papers, so it might be better to keep the reproducibility?

kan-bayashi replied:
OK.

kan-bayashi commented on egs/arctic/vc/local/text:
```diff
@@ -0,0 +1 @@
+../../../ljspeech/tts1/local/text
```
Now `local/text` is integrated with `espnet_tts_frontend`.
We do not need to use it (Maybe it does not exist).

kan-bayashi commented on espnet/nets/pytorch_backend/tacotron2/encoder.py:
```diff
@@ -162,3 +162,127 @@ def inference(self, x):
+            eunits (int, optional) The number of encoder blstm units.
+            econv_layers (int, optional) The number of encoder conv layers.
+            econv_filts (int, optional) The number of encoder conv filter size.
+            econv_chans (int, optional) The number of encoder conv filter channels.
```
Add docstring of `input_layer`.

kan-bayashi commented on espnet/nets/pytorch_backend/tacotron2/encoder.py:
```diff
@@ -162,3 +162,127 @@ def inference(self, x):
+            LongTensor: Batch of lengths of each sequence (B,)
+
+        """
+        if self.input_layer:
```
```suggestion
        if self.input_layer is not None:
```

kan-bayashi commented on espnet/nets/pytorch_backend/transformer/subsampling.py:
```diff
@@ -20,9 +20,59 @@ class Conv2dSubsampling(torch.nn.Module):
+
+    def __init__(self, idim, odim, dropout_rate, pos_enc=None):
+        """Construct an Conv2dSubsampling object."""
+        super(Conv2dSubsamplingCustomPosEnc, self).__init__()
```
Is it needed to separate module?
It can be implemented with options.
I think if you set default value and keep the same behavior when the option is default, it is OK.

kan-bayashi commented on espnet/nets/pytorch_backend/vc_tacotron2.py:
```diff
@@ -0,0 +1,755 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+# Copyright 2018 Nagoya University (Tomoki Hayashi)
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
```
Update.

kan-bayashi commented on espnet/nets/pytorch_backend/vc_tacotron2.py:
```diff
@@ -0,0 +1,755 @@
+            bce_pos_weight (float): Weight of positive sample of stop token.
+
+        """
+        super(Tacotron2Loss, self).__init__()
+        self.use_masking = use_masking
+        self.bce_pos_weight = bce_pos_weight
+
+    def forward(self, after_outs, before_outs, logits, ys, labels, olens):
+        """Calculate forward propagation.
+
+        Args:
+            after_outs (Tensor): Batch of outputs after postnets (B, Lmax, odim).
+            before_outs (Tensor): Batch of outputs before postnets (B, Lmax, odim).
+            logits (Tensor): Batch of stop logits (B, Lmax).
+            ys (Tensor): Batch of padded target features (B, Lmax, odim).
+            labels (LongTensor): Batch of the sequences of stop token labels (B, Lmax).
+            olens (LongTensor): Batch of the lengths of each target (B,).
+
+        Returns:
+            Tensor: L1 loss value.
+            Tensor: Mean square error loss value.
+            Tensor: Binary cross entropy loss value.
+
+        """
+        # perform masking for padded values
+        if self.use_masking:
+            mask = make_non_pad_mask(olens).unsqueeze(-1).to(ys.device)
+            ys = ys.masked_select(mask)
+            after_outs = after_outs.masked_select(mask)
+            before_outs = before_outs.masked_select(mask)
+            labels = labels.masked_select(mask[:, :, 0])
+            logits = logits.masked_select(mask[:, :, 0])
+
+        # calculate loss
+        l1_loss = F.l1_loss(after_outs, ys) + F.l1_loss(before_outs, ys)
+        mse_loss = F.mse_loss(after_outs, ys) + F.mse_loss(before_outs, ys)
+        bce_loss = F.binary_cross_entropy_with_logits(
+            logits, labels, pos_weight=torch.tensor(self.bce_pos_weight, device=ys.device))
+
+        return l1_loss, mse_loss, bce_loss
```
What is the difference campared to the TTS one?

kan-bayashi commented on espnet/nets/pytorch_backend/vc_tacotron2.py:
```diff
@@ -0,0 +1,755 @@
+from espnet.utils.fill_missing_args import fill_missing_args
+
+
+class GuidedAttentionLoss(torch.nn.Module):
+    """Guided attention loss function module.
+
+    This module calculates the guided attention loss described in `Efficiently Trainable Text-to-Speech System Based
+    on Deep Convolutional Networks with Guided Attention`_, which forces the attention to be diagonal.
+
+    .. _`Efficiently Trainable Text-to-Speech System Based on Deep Convolutional Networks with Guided Attention`:
+        https://arxiv.org/abs/1710.08969
+
+    """
```
What is the difference compared to TTS one?

kan-bayashi commented on espnet/nets/pytorch_backend/tacotron2/encoder.py:
```diff
@@ -162,3 +162,127 @@ def inference(self, x):
         return self.forward(xs, ilens)[0][0]
+
+
+class EncoderNoEmb(torch.nn.Module):
```
If the difference is only the input layer, I think it is not necessary to separate the module.

kan-bayashi commented on espnet/nets/pytorch_backend/vc_transformer.py:
```diff
@@ -0,0 +1,1083 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+# Copyright 2019 Tomoki Hayashi
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
```
Update

unilight commented:
Sorry for the late fix. (I was busy with SLT experiments lately...)
I have fixed many things according to the comments, refactored more, and added some documentation.
Please review this PR again if you feel needed.
Thanks!

sw005320 commented on espnet/nets/pytorch_backend/transformer/encoder.py:
```diff
@@ -25,7 +25,10 @@
     PositionwiseFeedForward,  # noqa: H301
```
FYI, with the conformer implementation, we have several variations for this subsampling.
We'll do some refactoring soon.

sw005320 commented on espnet/nets/pytorch_backend/transformer/subsampling.py:
```diff
@@ -17,21 +17,26 @@ class Conv2dSubsampling(torch.nn.Module):
     :param int idim: input dim
```
@pengchengguo, could you give @unilight your comments?
This is what we wanted to do?

pengchengguo replied:
Yes. Hi @unilight, we also need a custom positional encoding type for conformer model. To keep consistent with transformer's code, we want to add a parameter `pos_enc` to class `Conv2dSubsampling`. By default, `pos_enc=None` and we use `PositionalEncoding`. For specific scenarios, we will pass a `torch.nn.Module` to select different types, such as `ScaledPositionalEncoding` or `RelPositoinalEncoding`.  I think your implementation is same as we discussed.

unilight replied:
Thank you for your explanation, @pengchengguo ! If my implementation is correct, I will keep it this way.

sw005320 commented on utils/make_pair_json.py:
```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python3
```
Could you add a test for this?

sw005320 commented on utils/mcd_calculate.py:
```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/env python3
```
ditto

unilight commented:
Thank you for the review, @sw005320 !
I have some questions about writing tests since I have never done it before. I guess that since the scripts I am testing are in `utils`, I should follow the examples in `test_utils`. I know how to write pytests but I don't know how to interact with the `.bats` in `test_utils`. Can you give me some hints?

sw005320 commented:
It is documented a little bit in https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#unit-testing
In your cases, since VC is an important function, I want to have both types of unit tests (one for https://github.com/espnet/espnet/tree/master/test and the other for https://github.com/espnet/espnet/tree/master/test_utils).
You could make them locally and test them in advance by following  https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#unit-testing
Also, feel free to add descriptions there in this PR, if you think some parts are not clear (Also, feel free to ask us about the clarifications!).

Ideally, we want to have an integration test for VC (https://github.com/espnet/espnet/blob/master/ci/test_integration.sh). Note that the integration test is to check the whole recipe is working (we don't check the performance because it is too small to get some performance). If we can make such a small recipe for VC, we may also do it.

sw005320 commented:
@unilight, we can add some tests in the later PR.
It’s up to you.
If you fix the conflict/CI/renaming, I can merge it.

unilight commented:
@sw005320 I have added tests for `vc_tacotron2.py` and `vc_transformer.py`, as well as the new utils scripts `mcd_calculate.py` and `make_pair_json.py`.

> Ideally, we want to have an integration test for VC (https://github.com/espnet/espnet/blob/master/ci/test_integration.sh). Note that the integration test is to check the whole recipe is working (we don't check the performance because it is too small to get some performance). If we can make such a small recipe for VC, we may also do it.

It is not so straightforward to make a small testing recipe for VC but I will work on this later.

Also, I don't know why but the CI check is working strangely. Something seems to be wrong with the installation of espnet, but I didn't touch anything related to it (only added some necessary packages for VC in `setup.py`). Could you check it?

sw005320 commented on ci/test_shell.sh:
```diff
@@ -11,7 +11,7 @@ if ! [ -x "$(command -v bats)" ]; then
 fi
 if ! [ -x "$(command -v shellcheck)" ]; then
     echo "=== install shellcheck ==="
-    wget https://storage.googleapis.com/shellcheck/shellcheck-stable.linux.x86_64.tar.xz
```
could you revert to this?
I don't know why but the problem of the spellcheck seems to come from this.
Another possibility is to fix all shell errors.
I just saw them, and we could easily fix them.

unilight commented:
@sw005320 I have tried to fix all the CI errors but the coverage. I think the coverage decrease comes mostly from `espnet/utils/io_utils.py` since there is no test for new functions I implemented in it.

sw005320 commented:
> @sw005320 I have tried to fix all the CI errors except for the coverage. I think the coverage decrease comes mostly from `espnet/utils/io_utils.py` since there is no test for new functions I implemented in it.

Many thanks!!!
This is no problem.

kan-bayashi commented on egs/arctic/vc/conf/train_pytorch_transformer.nopt.n80.single.yaml:
```diff
@@ -0,0 +1 @@
+tuning/train_pytorch_transformer.nopt.v1.1.n80.single.yaml
```
How about making only the link to the best config? (We can keep `conf/tuning` as is)

unilight replied:
Do you mean to remove the `nopt` ones (without pre-training)? My original intention was to let users train models even if they don't want to use the pre-training technique (so they can fully reproduce the results in the paper). If you feel we can keep the best setting (i.e. TTS pre-training with the complete 1-hour data) I can do it.

kan-bayashi replied:
Thank you for your explanation.
Then, we can keep the links to `nopt` and `tts_pt` config.
I think `n80` is optional but if you want to keep as is, it is OK.

sw005320 commented:
I think this is should be merged soon after you finish renaming.

kan-bayashi commented:
Now I'm reviewing. Sorry for rate.

sw005320 commented:
> Now I'm reviewing. Sorry for rate.

No problem.
You actually reviewed this already, and I just forgot to check the interactions since it was 25 days ago.

kan-bayashi commented on espnet/nets/pytorch_backend/tacotron2/encoder.py:
```diff
@@ -52,6 +54,8 @@ def __init__(
 
         Args:
             idim (int) Dimension of the inputs.
+            use_embed (bool, optional) Whether to use embedding table or not.
```
Why don't you follow transformer encoder manner?
https://github.com/espnet/espnet/blob/71771245f9ea42a7e8c9ac619c47aa200cc33f6f/espnet/nets/pytorch_backend/transformer/encoder.py#L97-L127

kan-bayashi commented on espnet/nets/pytorch_backend/tacotron2/encoder.py:
```diff
@@ -52,6 +54,8 @@ def __init__(
         Args:
             idim (int) Dimension of the inputs.
+            use_embed (bool, optional) Whether to use embedding table or not.
+            encoder_reduction_factor (int, optional) Encoder reduction factor.
```
I think this is not needed since it can be represented by idim.

kan-bayashi commented on espnet/nets/pytorch_backend/transformer/subsampling.py:
```diff
@@ -17,21 +17,26 @@ class Conv2dSubsampling(torch.nn.Module):
+    def __init__(self, idim, odim, dropout_rate, pos_enc=None):
         """Construct an Conv2dSubsampling object."""
         super(Conv2dSubsampling, self).__init__()
+        if pos_enc:
```
```suggestion
        if pos_enc is not None:
        	assert isinstance(pos_enc, torch.nn.Module)
```

kan-bayashi commented on espnet/nets/pytorch_backend/transformer/subsampling.py:
```diff
@@ -17,21 +17,26 @@ class Conv2dSubsampling(torch.nn.Module):
         self.out = torch.nn.Sequential(
-            torch.nn.Linear(odim * (((idim - 1) // 2 - 1) // 2), odim),
-            PositionalEncoding(odim, dropout_rate),
+            torch.nn.Linear(odim * (((idim - 1) // 2 - 1) // 2), odim), self.pos_enc
```
Is it need to register `pos_enc` in `self`?

kan-bayashi commented on espnet/nets/pytorch_backend/vc_tacotron2.py:
```diff
@@ -0,0 +1,785 @@
+
+        # define network modules
+        self.enc = Encoder(
+            idim=idim,
+            use_embed=False,
+            encoder_reduction_factor=args.encoder_reduction_factor,
```
Why don't you change here `idim=idim + args.encoder_reduction_factor`?

unilight commented:
@sw005320 @kan-bayashi I reflected the reviews, and the renaming is done. I've also updated some to-dos for future PRs. I think this PR is ready to go!

sw005320 commented:
Super great job @unilight !!!
